### ISSUE TITLE
Custom pdf export - user display name on top of sheet

### ISSUE THREAD

**Issue by Recoil1980** [kevinpapst/kimai2#1931]
Currently I am trying to edit one of the templates for exporting the weeks hours for a user to a pdf.  From the dashboard, I go to Export, then set the Time Range and User.

Above the Period date I have:
`
<h2 style="margin-bottom: 0; padding-bottom: 0">{{ kimai_context.branding.company|raw }}</h2>
<h3 style="margin-bottom: 0; padding-bottom: 0">
    {{ 'label.user'|trans }}: {{ query.user.displayName }}
</h3>
`

It is showing the company name and `User:` label, but not the selected users name.  I am at a loss on exactly what to put instead of `query.user.displayName`.  There's not much information on [exporting](https://www.kimai.org/documentation/export.html#adding-export-templates) to suggest what I'm looking for.

**Comment by kevinpapst** [comment]
Export has multi-select for users, so '{{ query.user }}` is always empty.
I guess you have to use `{{ query.users.0.displayName }}`

**Comment by Recoil1980** [comment]
Thanks so much!  That did the trick, I'll just have to remember to only export 1 users hours at a time.  Trying to export 2 of them is only going to show the first users name.

I will likely add another plugin request to my list to make the user select field single select.  Not sure if this is possible at the moment...

**Comment by kevinpapst** [comment]
This is definitely not wanted and will be rejected. It took me quite a while to get to the point that almost all fields can be multi-select.

**Comment by Recoil1980** [comment]
I understand your point.  Would it be possible on the multi-select to pay for a custom template on the `Summary` and `Full List` for all the selected users separately?

Our work days go from Friday to the following Thursday, so I will have to select those days for each user individually to export.  The only benefit is saving time really, but at some point someone else besides myself may end up using this system.  Honestly it isn't that difficult to remove one user from the multi-select, add a different one, and export a separate sheet to print.

**Comment by kevinpapst** [comment]
New exports can be added via custom plugin => so yes, a custom template is possible.

**Comment by Recoil1980** [comment]
Is there a list of how to access the data anywhere?  I would have never guessed how to get the users display name, but going off that logic, `query.users.0.hourly_rate` or `query.users.0.hourlyRate` or `query.users.0.rate` or `query.users.0.rate_hourly` should return that value for that user, but none of them are working...

I hate to keep asking tedious questions, so if there is a reference it would help.

**Comment by kevinpapst** [comment]
Switch to dev mode (APP_ENV=DEv in .env) and then add a `{{ dump(query) }}` and see all available fields immediately.

All available fields for the user can be seen here: https://github.com/kevinpapst/kimai2/blob/master/src/Entity/User.php (search for `public function get`). Then a function like `public function getLanguage()` is available in the template with `user.language`.

There is magic involved that is already documented here: https://twig.symfony.com

The hourly rate is a special field, which has no own accessor method, as it is used in only 1 or 2 places in Kimai.
You can grab it with `{{ user.preferenceValue('hourly_rate') }}` => calling the function `getPreferenceValue()`.

**Comment by limitedAtonement** [comment]
I tried adding a `{{ dump }}` and it's not working.

I have tag 1.13 checked out, and my `.env` says `APP_ENV=dev`. In `templates/timesheet/index.html.twig` after line 9, I added `{{ dump(tableName) }}`. When I run the application and go to http://127.0.0.1:8080/en/timesheet/, I get "Something is wrong" in the browser, my server shows that it returned a 500 status code, and `var/log/prod.log` says,

```
[2021-02-09 16:36:12] request.CRITICAL: Uncaught PHP Exception Twig\Error\SyntaxError: "Did you forget to run "composer require symfony/debug-bundle"? Unknown function "dump" in "timesheet/index.html.twig"." at /app/vendor/symfony/twig-bridge/UndefinedCallableHandler.php line 96 {"exception":"[object] (Twig\\Error\\SyntaxError(code: 0): Did you forget to run \"composer require symfony/debug-bundle\"? Unknown function \"dump\" in \"timesheet/index.html.twig\". at /app/vendor/symfony/twig-bridge/UndefinedCallableHandler.php:96)"} []
[2021-02-09 16:36:12] security.DEBUG: Stored the security token in the session. {"key":"_security_secured_area"} []
```
I ran `composer require symfony/debug-bundle;` as requested, but that didn't do any good (it looks like `symfony/debug-bundle` is already installed in composer.json).

Did something change since August that would cause this not to work, or am I doing something wrong? Thank you!

**Comment by kevinpapst** [comment]
You need to run `composer install` to get the development packages. Then it should work.

**Comment by limitedAtonement** [comment]
I had run that. Running it now shows `Nothing to install, update or remove`:

```
bash-5.0# composer install
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Package doctrine/reflection is abandoned, you should avoid using it. Use roave/better-reflection instead.
Package zendframework/zend-escaper is abandoned, you should avoid using it. Use laminas/laminas-escaper instead.
Package fzaninotto/faker is abandoned, you should avoid using it. No replacement was suggested.
Package phpunit/php-token-stream is abandoned, you should avoid using it. No replacement was suggested.
Generating autoload files
composer/package-versions-deprecated: Generating version class...
composer/package-versions-deprecated: ...done generating version class
132 packages you are using are looking for funding.
Use the `composer fund` command to find out more!

Synchronizing package.json with PHP packages
Don't forget to run npm install --force or yarn install --force to refresh your JavaScript dependencies!
Run composer recipes at any time to see the status of your Symfony recipes.

Executing script cache:clear [OK]
Executing script assets:install [OK]

bash-5.0#
```

(and I get the same error when visiting the URL mentioned above.)

**Comment by kevinpapst** [comment]
Then you are not running in APP_ENV=dev.
Check the Doctor screen.In the title there need to be "Doctor Environment: dev - Version: 1.13 stable"

**Comment by limitedAtonement** [comment]
My thoughts exactly. Thank you for pointing out where I can see the environment. On my Doctor screen, I see `Environment: prod – Version: 1.13 stable`.

I have set the APP_ENV environment variable as well as making sure APP_ENV in the `.env` file is set to dev. To convince myself that the environment is correct, I adjusted my launch script to `echo environment is $APP_ENV;` before `composer install;` and before launching the web server, and it says "dev" both times. In the DoctorController constructor, I added `error_log("env is " . var_export($_ENV, true));`, and it shows that APP_ENV is `dev` as well. I'm convinced that PHP's process environment is as desired, but let me know if something still doesn't look right.

In `DoctorController`, of course, `error_log("environment is " . var_export($kernelEnvironment,true));` shows `prod`. I did a bit of digging to see whence this `$kernelEnvironment` comes. It appears to be set in Symfony's http Kernel object (the base object of our src/Kernel.php), but I don't know what calls that constructor.

What else should I be looking for to get the application into debug mode?
